src/EditorFeatures/Core/ExternalAccess/IntelliCode/Api/IIntentProcessor.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.IntelliCode.Api
+{
+    internal interface IIntentProcessor
+    {
+        /// <summary>
+        /// For an input intent, computes the edits required to apply that intent and returns them.
+        /// </summary>
+        /// <param name="intentRequestContext">the intents with the context in which the intent was found.</param>
+        /// <returns>the edits that should be applied to the current snapshot.</returns>
+        Task<ImmutableArray<IntentResult>> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Defines the data needed to compute the code action edits from an intent.
+    /// </summary>
+    internal struct IntentRequestContext
+    {
+        /// <summary>
+        /// The intent name.  <see cref="WellKnownIntents"/> contains all intents roslyn knows how to handle.
+        /// </summary>
+        public string IntentName { get; }
+
+        /// <summary>
+        /// JSON formatted data specific to the intent that must be deserialized into the appropriate object.
+        /// </summary>
+        public string? IntentData { get; }
+
+        /// <summary>
+        /// The text snapshot and selection when <see cref="IIntentProcessor.ComputeEditsAsync"/>
+        /// was called to compute the text edits and against which the resulting text edits will be calculated.
+        /// </summary>
+        public SnapshotSpan CurrentSnapshotSpan { get; }
+
+        /// <summary>
+        /// The text edits that should be applied to the <see cref="CurrentSnapshotSpan"/> to calculate
+        /// a prior text snapshot before the intent happened.  The snapshot is used to calculate the actions.
+        /// </summary>
+        public ImmutableArray<TextChange> PriorTextEdits { get; }
```

Comment:
@CyrusNajmabadi currently this is using roslyn text change types as there is no editor text change type that can be constructed from just a span + new text.  So  I think the only other option here is LSP.  Not sure which is preferable.

Reply:
i'm not sure the best way to state it but it's worth callign out that this is a set of edits that is applied at once, not linerally in a row (or is it the latter?).  in other words, you don't apply one edit, get a new snapshot, then apply teh next edit to it.  you apply the full set to get the original snapshot.